otredis.go

```diff
@@ -2,34 +2,77 @@ package otredis
 
 import (
 	"context"
+	"fmt"
+	"strings"
 
 	"github.com/go-redis/redis"
-	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
 )
 
-// WrapRedisClient adds opentracing measurements for commands and returns cloned client
-func WrapRedisClient(ctx context.Context, c *redis.Client) *redis.Client {
+func WrapRedisClient(ctx context.Context, client *redis.Client) *redis.Client {
 	if ctx == nil {
-		return c
+		return client
 	}
 	parentSpan := opentracing.SpanFromContext(ctx)
 	if parentSpan == nil {
-		return c
+		return client
 	}
+	ctxClient := client.WithContext(ctx)
+	opts := ctxClient.Options()
+	ctxClient.WrapProcess(process(parentSpan, opts))
+	ctxClient.WrapProcessPipeline(processPipeline(parentSpan, opts))
+	return ctxClient
+}
 
-	// clone using context
-	copy := c.WithContext(c.Context())
-	copy.WrapProcess(func(oldProcess func(cmd redis.Cmder) error) func(cmd redis.Cmder) error {
+func process(parentSpan opentracing.Span, opts *redis.Options) func(oldProcess func(cmd redis.Cmder) error) func(cmd redis.Cmder) error {
+	return func(oldProcess func(cmd redis.Cmder) error) func(cmd redis.Cmder) error {
 		return func(cmd redis.Cmder) error {
-			tr := parentSpan.Tracer()
-			sp := tr.StartSpan("redis", opentracing.ChildOf(parentSpan.Context()))
-			ext.DBType.Set(sp, "redis")
-			sp.SetTag("db.method", cmd.Name())
-			defer sp.Finish()
-
+			dbMethod, dbStatement := formatCommandAsDbTags(cmd)
+			doSpan(parentSpan, opts, "redis-cmd", dbMethod, dbStatement)
 			return oldProcess(cmd)
 		}
-	})
-	return copy
+	}
+}
+
+func processPipeline(parentSpan opentracing.Span, opts *redis.Options) func(oldProcess func(cmds []redis.Cmder) error) func(cmds []redis.Cmder) error {
+	return func(oldProcess func(cmds []redis.Cmder) error) func(cmds []redis.Cmder) error {
+		return func(cmds []redis.Cmder) error {
+			dbMethod, dbStatement := formatCommandsAsDbTags(cmds)
+			doSpan(parentSpan, opts, "redis-pipeline-cmd", dbMethod, dbStatement)
+			return oldProcess(cmds)
+		}
+	}
+}
+
+func formatCommandAsDbTags(cmd redis.Cmder) (string, string) {
+	dbMethod := cmd.Name()
+	var sprintArgs []string
+	for _, arg := range cmd.Args() {
+		sprintArgs = append(sprintArgs, fmt.Sprint(arg))
```

Comment:
it's not a good idea to put a value of the commands into tags.

For some use cases, the value of `set` can be huge (megabytes of data, sometimes gzipped).
Another case is when the value contains sensitive information that shouldn't be exposed to monitoring (like some password or credit card number). Actually, even a key may contain sensitive information. So it might be not safe for `get` too.

Reply:
We have some kind of requirement about that but it is safer as you explain, no prob

Reply:
If you have a use case that requires setting arguments into tags we can add it in a separate pull request (but disabled by default).